src/extension/prompt/node/repoInfoTelemetry.ts

```diff
@@ -0,0 +1,194 @@
+	private _beginTelemetryPromise: Promise<void> | undefined;
+
+	constructor(
+		private readonly _telemetryMessageId: string,
+		@ITelemetryService private readonly _telemetryService: ITelemetryService,
+		@IGitService private readonly _gitService: IGitService,
+		@IGitDiffService private readonly _gitDiffService: IGitDiffService,
+		@IGitExtensionService private readonly _gitExtensionService: IGitExtensionService,
+		@ICopilotTokenStore private readonly _copilotTokenStore: ICopilotTokenStore,
+		@ILogService private readonly _logService: ILogService,
+		@IFileSystemService private readonly _fileSystemService: IFileSystemService,
+	) { }
+
+	/*
+	* Sends the begin event telemetry
+	*/
+	public async sendBeginTelemetryIfNeeded(): Promise<void> {
+		if (this._beginTelemetrySent) {
+			// Already sent or in progress
+			return this._beginTelemetryPromise;
+		}
+
+		this._beginTelemetrySent = true;
+		this._beginTelemetryPromise = this._sendRepoInfoTelemetry('begin');
+
+		return this._beginTelemetryPromise.catch((error) => {
+			this._logService.warn(`Failed to send begin repo info telemetry ${error}`);
+		});
+	}
+
+	/*
+	* Sends the end event telemetry
+	*/
+	public async sendEndTelemetry(): Promise<void> {
+		await this._beginTelemetryPromise;
+
+		return this._sendRepoInfoTelemetry('end').catch((error) => {
+			this._logService.warn(`Failed to send end repo info telemetry ${error}`);
+		});
+	}
+
+	private async _sendRepoInfoTelemetry(location: 'begin' | 'end'): Promise<void> {
+		if (this._copilotTokenStore.copilotToken?.isInternal !== true) {
+			return;
+		}
+
+		const gitInfo = await this._getRepoInfoTelemetry();
+		if (!gitInfo) {
+			return;
+		}
+
+		// Multiplex properties will split up the large properties (diffsJSON) into multiple properties
+		// that we can combine later.
+		const properties = multiplexProperties({
+			...gitInfo,
+			location,
+			telemetryMessageId: this._telemetryMessageId
+		} as RepoInfoInternalTelemetryProperties);
+
+		this._telemetryService.sendInternalMSFTTelemetryEvent('request.repoInfo', properties);
+	}
+
+	private async _getRepoInfoTelemetry(): Promise<RepoInfoTelemetryProperties | undefined> {
+		const repoContext = this._gitService.activeRepository.get();
+
+		if (!repoContext || !repoContext.changes) {
+			return;
+		}
+
+		const githubInfo = getGitHubRepoInfoFromContext(repoContext);
+		if (!githubInfo) {
+			return;
+		}
+
+		// Get the upstream commit from the repository
+		const gitAPI = this._gitExtensionService.getExtensionApi();
+		const repository = gitAPI?.getRepository(repoContext.rootUri);
+		const upstreamCommit = repository?.state.HEAD?.upstream?.commit;
+		if (!upstreamCommit) {
+			return;
+		}
+
+		// Before we calculate our async diffs, sign up for file system change events
+		// Any changes during the async operations will invalidate our diff data and we send it
+		// as a failure without a diffs
+		const watcher = this._fileSystemService.createFileSystemWatcher('**/*');
```

Comment:
Per teams discussion, if files are changed during the snapshot process just reject the event. We'll log that in telemetry so if this is happening more often than we expect we can revisit alternative ways to handle this. However this current approach seems the best way to start, as other options might have to await on some amount of data fetching. The core issue of wanting to not block, but capturing against a live git repository state feels like a trade off one way or another and the tradeoff here to be non-blocking and just reject some % of events seems right.